MIAL Super-Resolution Toolkit v2.0

sebastientourbier opened the pull request:
This PR is dedicated to the upcoming MIAL Super-Resolution Toolkit v2.0 release which will adopt the BIDS standard and will provide a workflow, implemented in Python using the Nipype library, that will interface with the MIALSRTK C++ tools and will be executed following the BIDS-App standard.

To be done before release:
- [x]  Creation of BIDS sample dataset
- [x]  Update CI and current example scripts with new BIDS organization
- [x]  Creation of `pymialsrtk` package with `setup.py`
-  Implementation of Nipype components interfacing with MIALSRTK C++ compiled tools
    - [x]  Implementation of all Nipype interfaces necessary to reproduce the pipeline implemented in the original bash example script,  along with the creation of dedicated example  jupyter notebooks for example and testing.
    - [x]  Implementation of Nipype workflow reproducing the pipeline with manual masks, along with the creation of dedicated example  jupyter notebooks for example and testing.
    - [x] Use of Nipype interface `DataSinker` to move and rename files according to BIDS derivatives standard 
- [x]  Creation of the BIDS-App `run.py` entrypoint script that should parse BIDS-App input flags and build workflows (multiple subjects/sessions) according those inputs. 
        with https://github.com/brainhack-ch/superres-mri/blob/master/sinapps/nlmdenoise/run.py as starting point. 
        *Discussed:* 
         how to input parameters specific to the SR pipeline such as list of stacks, stack order, TV parameters, ... (see https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/pull/2#issuecomment-637673761)
- [x]  Review Dockerfile for the BIDS-App 
- [x]  Review CI:  build, test and deployment of two different  docker images (one encapsulating the MIALSRTK compiled C++ tools, one specifically dedicated to the BIDS App)
- [x]  Integration of DL-based brain localization and extraction interface in workflow for automatic pipeline (merge Hamza branch, modify parser and add the interface in the workflow implemented in run.py) - See PR #4 
- [x] Review output structure for multiple runs with different list/order of scans
- Documentation 
  - [x] Code documentation with docstrings (https://www.python.org/dev/peps/pep-0257/)
  - [x] Update ``documentation/outputs.rst`` and ``documentation/citing.rst`` in software documentation on readthedocs (rendered @ https://mialsrtk.readthedocs.io/en/dev-pgd-hk/). In particular:
- [x] Improve code quality based on codacy report: https://app.codacy.com/gh/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/pullRequest?prid=6104929
- [x] Last review before merge

sebastientourbier commented:
For the moment this provides the complete solution to run the SR pipeline with manual masks as inputs. 

Should we also integrate your implementation for automated DL-based brain localization and extraction? @pdedumast @hamzake
I guess this PR already includes a number of major changes. This could come later in an next 2.1 minor release version.

sebastientourbier commented on pymialsrtk/info.py:
```diff
@@ -0,0 +1,36 @@
+_version_micro = 0
+_version_extra = '-beta-20190906'
+
+__release_date__ = '06.09.2019'
```
To be updated when we know

sebastientourbier commented on pymialsrtk/interfaces/utils.py:
```diff
@@ -0,0 +1,16 @@
+# Copyright © 2016-2019 Medical Image Analysis Laboratory, University Hospital Center and University of Lausanne (UNIL-CHUV), Switzerland
```
Update copyright

sebastientourbier commented on pymialsrtk/interfaces/reconstruction.py:
```diff
@@ -0,0 +1,311 @@
+# Copyright © 2016-2019 Medical Image Analysis Laboratory, University Hospital Center and University of Lausanne (UNIL-CHUV), Switzerland
```
Update copyright

sebastientourbier commented on pymialsrtk/interfaces/postprocess.py:
```diff
@@ -0,0 +1,171 @@
+# Copyright © 2016-2019 Medical Image Analysis Laboratory, University Hospital Center and University of Lausanne (UNIL-CHUV), Switzerland
```
Update copyright

sebastientourbier commented on pymialsrtk/interfaces/preprocess.py:
```diff
@@ -0,0 +1,595 @@
+# Copyright © 2016-2019 Medical Image Analysis Laboratory, University Hospital Center and University of Lausanne (UNIL-CHUV), Switzerland
```
Update copyright

sebastientourbier commented on scripts/superresolution:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env python
```
Not sure if this script is still necessary as we will implement the nipype pipeline and BIDS app parser in `docker/bidsapp/run.py`

sebastientourbier commented:
Here comes an example of how a json config file (input to the BIDS app config_file) could look like (assuming a simple subject structure - no session):
```
{
  "01": {
    "ordered-stack-list": [1, 2, 4, 5, 3, 6],
    "lambdaTV": 0.05,
    "deltatTV": 0.1
  },
  
  "02": {
    "ordered-stack-list": [2, 1, 4, 5, 6, 3],
    "lambdaTV": 0.05,
    "deltatTV": 0.1
  }
}
```

We could think about having one config file (let's say `pymialsrtk_config.json`) in the `code` directory of the BIDS dataset to be processed, that could store the parameters for ALL participants (cases) and then in the BIDS app entrypoint `run.py`, we extract from the json file parameters corresponding to the case(s) specified by the list of subject labels to be processed.

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+
+## Node linkage
+## Node linkage
+def create_workflow(bids_dir, process_dir, subject, p_stacksOrder, session=None, paramTV={}, srID=None):
```
I would change `process_dir` back to `output_dir` to be more compliant to BIDS and BIDS App and much more flexible. This instance this can allow us to read inputs from a dataset stored remotely and mounted locally (specified by `bids_dir`) and saving results in and output directory locally (specified by `output_dir`).

We can think about having a data derivative structure like:
```
output_dir/
  |____ run-[srID]/
            |____ pymialsrtk/
            |         |____ sub-01/
            |                  |____ anat/
            |____ nipype/
                      |____ sub-01/
                                |____ anatomical_pipeline/

```
with:
* `process_dir` set to 'output_dir/run-[srID]/'
* `wf_base_dir` set to 'output_dir/run-[srID]/nipype/sub-01/anatomical_pipeline/'
* `datasink.inputs.base_directory` set to `output_dir/run-[srID]/pymialsrtk/sub-01/anatomical_pipeline/`

I would not use `mialsrtk-py` as this is planned in BIDS derivatives extension to use the convention `<pipeline>-<variant>` (see https://bids-specification.readthedocs.io/en/derivatives/05-derivatives/01-introduction.html). In this future we could also adopt this convention. 
This would be quite straight forward as we can access it with the following code: `from pymialsrtk.info import __version__`.

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+
+    
+
+    if session is None:
+        if srID is None:
+            wf_base_dir = os.path.join(process_dir, subject)
+            process_dir = os.path.join(process_dir, subject)
+        else:
+            wf_base_dir = os.path.join(process_dir, subject, str(srID))
+            process_dir = os.path.join(process_dir, subject, str(srID))
+    else:
+        if srID is None:
+            wf_base_dir = os.path.join(process_dir, subject, session)
+            process_dir = os.path.join(process_dir, subject, session)
+        else:
+            wf_base_dir = os.path.join(process_dir, subject, session, str(srID))
+            process_dir = os.path.join(process_dir, subject, session, str(srID))
```
Would need to be changed related to first comment

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+            wf_base_dir = os.path.join(process_dir, subject, session, str(srID))
+            process_dir = os.path.join(process_dir, subject, session, str(srID))
+
+    if not os.path.exists(process_dir):
+        os.makedirs(process_dir)
+    print("Process directory: {}".format(wf_base_dir))
```
Might need to be updated regarding changes related to first comment

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+
+
+    datasink = Node(DataSink(), name='sinker')
+    output_dir = os.path.join("{}".format(bids_dir),"derivatives","mialsrtk-py", subject)
+    if session is not None:
+        output_dir = os.path.join(output_dir, session)
+    datasink.inputs.base_directory = output_dir
+    
```
Would need to be changed related to first comment

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+    dictsink = JSONFileSink(name='jsonsinker')
+    dictsink.inputs.in_dict = output_dict
+
+    if srID is None:
+        dictsink.inputs.out_file = os.path.join(output_dir, 'anat', sub_ses+'_rec-SR_T2w.json')
+    else:
+        dictsink.inputs.out_file = os.path.join(output_dir, 'anat', sub_ses+'_rec-SR'+'_srID-'+str(int(srID))+'_T2w.json')  
```
Would need to be changed related to first comment

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+                        print('Do not process subjects %s because of missing parameters.' % sub)
+                        continue
+
+                    if 'paramTV' in sr_params.keys():
+                        if 'srID' in sr_params.keys():
+                            main(bids_dir=args.bids_dir, 
+                                process_dir=args.output_dir, 
+                                subject=sub, 
+                                p_stacksOrder=sr_params['stacksOrder'], 
+                                session=ses, 
+                                paramTV=sr_params['paramTV'], 
+                                srID=sr_params['srID'])
+                        else:
+                            main(bids_dir=args.bids_dir, 
+                                process_dir=args.output_dir, 
+                                subject=sub, 
+                                p_stacksOrder=sr_params['stacksOrder'], 
+                                session=ses, 
+                                paramTV=sr_params['paramTV'])
+
+                    else:
+                        if 'srID' in sr_params.keys(): 
+                            main(bids_dir=args.bids_dir, 
+                                process_dir=args.output_dir, 
+                                subject=sub, 
+                                p_stacksOrder=sr_params['stacksOrder'], 
+                                session=ses, 
+                                srID=sr_params['srID'])
+                        else:
+                            main(bids_dir=args.bids_dir, 
+                                process_dir=args.output_dir, 
+                                subject=sub, 
+                                p_stacksOrder=sr_params['stacksOrder'], 
+                                session=ses)
```
Related to first comment: would need to update `process_dir` to `output_dir`

sebastientourbier commented:
Once output structure has been corrected, we could then integrate the brain localization module of @hamzake. For this, it would require:
* Finalize pull request (PR) #4 that would merge hamza branch into the branch dedicated to this PR.
* move `Network_checkpoints`  folder from `data` to `pymialsrtk/data`.

pdedumast commented:
> @pdedumast Here comes an example of how a json config file (input to the BIDS app config_file) could look like (assuming a simple subject structure - no session):
> 
> ```
> {
>   "01": {
>     "ordered-stack-list": [1, 2, 4, 5, 3, 6],
>     "lambdaTV": 0.05,
>     "deltatTV": 0.1
>   },
>   
>   "02": {
>     "ordered-stack-list": [2, 1, 4, 5, 6, 3],
>     "lambdaTV": 0.05,
>     "deltatTV": 0.1
>   }
> }
> ```
> 
> We could think about having one config file (let's say `pymialsrtk_config.json`) in the `code` directory of the BIDS dataset to be processed, that could store the parameters for ALL participants (cases) and then in the BIDS app entrypoint `run.py`, we extract from the json file parameters corresponding to the case(s) specified by the list of subject labels to be processed.


@sebastientourbier, here is what is currently implemented in the last commit:
```
{
  "01": [
    { "sr-id":"001",
      "stacksOrder": [1, 2, 4, 5, 3, 6],
      "paramTV": { 
        "lambdaTV": 0.75, 
        "deltatTV": 0.01,
        "primal_dual_loops": 10  }
    },
    { 
      "sr-id":"002",
      "session":"001",
      "stacksOrder": [1, 5, 4]
    }]
}
```
A list of reconstruction can be specified for a subject, which allows to proceed to the reconstruction of the same exam with different parameters (identified with `sr-id`) or of different MR exams sessions. 
`session` and `paramTV` are not mandatory entries.

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+        print( 'SRTV_'+sub_ses+'_'+str(len(p_stacksOrder))+'V_rad1_gbcorr.nii.gz', '    --->     ', sub_ses+'_rec-SR_T2w.nii.gz')
+        substitutions.append( ( 'SRTV_'+sub_ses+'_'+str(len(p_stacksOrder))+'V_rad1_gbcorr.nii.gz', sub_ses+'_rec-SR_T2w.nii.gz') )
+        
+        print( sub_ses+'_T2w_uni_bcorr_histnorm_srMask.nii.gz', '    --->     ', sub_ses+'_rec-SR_T2w_desc-brain_mask.nii.gz')
+        substitutions.append( ( sub_ses+'_T2w_uni_bcorr_histnorm_srMask.nii.gz', sub_ses+'_rec-SR_T2w_desc-SRmask.nii.gz') )
```
Typo: I guess it should be `_rec-SR_T2w_desc-brain_mask.nii.gz', not `_rec-SR_T2w_desc-SRmask.nii.gz`

sebastientourbier replied:
Correction:
`sub_ses+'_rec-SR_T2w_desc-brain_mask.nii.gz'`-> `sub_ses+'_rec-SR_desc-brain_mask.nii.gz'`

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+    if srID is None:
+        for stack in p_stacksOrder:
+        
+            print( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm.nii.gz', '    --->     ',sub_ses+'_run-'+str(stack)+'_T2w_preproc.nii.gz')
+            substitutions.append( ( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm.nii.gz', sub_ses+'_run-'+str(stack)+'_T2w_preproc.nii.gz') )
```
Correction:
`sub_ses+'_run-'+str(stack)+'_T2w_preproc.nii.gz'` -> `sub_ses+'_run-'+str(stack)+'_desc-preproc_T2w.nii.gz`

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+            print( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm_transform_'+str(len(p_stacksOrder))+'V.txt', '    --->     ', sub_ses+'_run-'+str(stack)+'_T2w_from-origin_to-SDI_mode-image_xfm.txt')
+            substitutions.append( ( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm_transform_'+str(len(p_stacksOrder))+'V.txt', sub_ses+'_run-'+str(stack)+'_T2w_from-origin_to-SDI_mode-image_xfm.txt') )
+            
+            print( sub_ses+'_run-'+str(stack)+'_T2w_uni_bcorr_histnorm_LRmask.nii.gz', '    --->     ', sub_ses+'_run-'+str(stack)+'_T2w_desc-LRmask.nii.gz')
+            substitutions.append( ( sub_ses+'_run-'+str(stack)+'_T2w_uni_bcorr_histnorm_LRmask.nii.gz', sub_ses+'_run-'+str(stack)+'_T2w_desc-LRmask.nii.gz') )
```
Correction:
`sub_ses+'_run-'+str(stack)+'_T2w_desc-LRmask.nii.gz'` -> `sub_ses+'_run-'+str(stack)+'_desc-brain_mask.nii.gz'`

sebastientourbier commented on pymialsrtk/pipelines/anatomical/srr.py:
```diff
@@ -0,0 +1,494 @@
+        substitutions.append( ( sub_ses+'_T2w_uni_bcorr_histnorm_srMask.nii.gz', sub_ses+'_rec-SR_T2w_desc-SRmask.nii.gz') )
+
+    else:
+        for stack in p_stacksOrder:
+        
+            print( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm.nii.gz', '    --->     ',sub_ses+'_run-'+str(stack)+'_id-'+srID+'_T2w_preproc.nii.gz')
+            substitutions.append( ( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm.nii.gz', sub_ses+'_run-'+str(stack)+'_id-'+srID+'_T2w_preproc.nii.gz') )
+            
+            print( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm_transform_'+str(len(p_stacksOrder))+'V.txt', '    --->     ', sub_ses+'_run-'+str(stack)+'_id-'+srID+'_T2w_from-origin_to-SDI_mode-image_xfm.txt')
+            substitutions.append( ( sub_ses+'_run-'+str(stack)+'_T2w_nlm_uni_bcorr_histnorm_transform_'+str(len(p_stacksOrder))+'V.txt', sub_ses+'_run-'+str(stack)+'_id-'+srID+'_T2w_from-origin_to-SDI_mode-image_xfm.txt') )
+            
+            print( sub_ses+'_run-'+str(stack)+'_T2w_uni_bcorr_histnorm_LRmask.nii.gz', '    --->     ', sub_ses+'_run-'+str(stack)+'_id-'+srID+'_T2w_desc-LRmask.nii.gz')
+            substitutions.append( ( sub_ses+'_run-'+str(stack)+'_T2w_uni_bcorr_histnorm_LRmask.nii.gz', sub_ses+'_run-'+str(stack)+'_id-'+srID+'_T2w_desc-LRmask.nii.gz') )
+
+            
+        print( 'SDI_'+sub_ses+'_'+str(len(p_stacksOrder))+'V_rad1.nii.gz', '    --->     ', sub_ses+'_rec-SDI'+'_id-'+srID+'_T2w.nii.gz')
+        substitutions.append( ( 'SDI_'+sub_ses+'_'+str(len(p_stacksOrder))+'V_rad1.nii.gz', sub_ses+'_rec-SDI'+'_id-'+srID+'_T2w.nii.gz') )
+
+        print( 'SRTV_'+sub_ses+'_'+str(len(p_stacksOrder))+'V_rad1_gbcorr.nii.gz', '    --->     ', sub_ses+'_rec-SR'+'_id-'+srID+'_T2w.nii.gz')
+        substitutions.append( ( 'SRTV_'+sub_ses+'_'+str(len(p_stacksOrder))+'V_rad1_gbcorr.nii.gz', sub_ses+'_rec-SR'+'_id-'+srID+'_T2w.nii.gz') )
+        
+        print( sub_ses+'_T2w_uni_bcorr_histnorm_srMask.nii.gz', '    --->     ', sub_ses+'_rec-SR'+'_id-'+srID+'_T2w_desc-brain_mask.nii.gz')
+        substitutions.append( ( sub_ses+'_T2w_uni_bcorr_histnorm_srMask.nii.gz', sub_ses+'_rec-SR'+'_id-'+srID+'_T2w_desc-SRmask.nii.gz') )
```
You would not have to differentiate in name has each of the `output_dir/run-<srID>` folder would correspond to an independent BIDS derivatives directory

sebastientourbier commented:
> @sebastientourbier, here is what is currently implemented in the last commit:
> 
> ```
> {
>   "01": [
>     { "sr-id":"001",
>       "stacksOrder": [1, 2, 4, 5, 3, 6],
>       "paramTV": { 
>         "lambdaTV": 0.75, 
>         "deltatTV": 0.01,
>         "primal_dual_loops": 10  }
>     },
>     { 
>       "sr-id":"002",
>       "session":"001",
>       "stacksOrder": [1, 5, 4]
>     }]
> }
> ```
> 
> A list of reconstruction can be specified for a subject, which allows to proceed to the reconstruction of the same exam with different parameters (identified with `sr-id`) or of different MR exams sessions.
> `session` and `paramTV` are not mandatory entries.

Tip top 👍

sebastientourbier commented:
These new commits involve:
1. the creation of the BIDS App (Dockerfile / pymialsrtk and dependencies installation)
2. the integration of circleci for continuous integration testing of the BIDS App
Also I had to make a few fixes and made the output more BIDS compliant (the question of `id-<srId>` and the naming of the different pipeline run - now 'rec<srId>' - is still opened.)

sebastientourbier commented:
Here is the preview of the documentation draft on readthedocs. Not sure it is publicly acciessible though:
https://mialsrtk.readthedocs.io/en/dev-pgd-hk/

sebastientourbier commented:
@all-contributors please add @sebastientourbier for design, tests and code

sebastientourbier commented:
@all-contributors please add @sebastientourbier  for examples, data, doc and review
@all-contributors please add @xbresson for code
@all-contributors please add @pdedumast for data, examples, tests and code
@all-contributors please add @hamzake  for data, examples, tests and code
@all-contributors please add @meribach  for fundingFinding

sebastientourbier commented:
@all-contributors please add @sebastientourbier for examples, data, doc and review

sebastientourbier commented:
@all-contributors please add @xbresson for code

sebastientourbier commented:
@all-contributors please add @pdedumast for data, examples, tests and code

sebastientourbier commented:
@all-contributors please add @hamzake for data, examples, tests and code

sebastientourbier commented:
@all-contributors please add @meribach for fundingFinding

sebastientourbier commented:
@all-contributors please add @pdedumast for data, examples, tests and code

sebastientourbier commented:
@all-contributors please add @hamzake for data, examples, tests and code

sebastientourbier commented:
@all-contributors please add @meribach for fundingFinding

sebastientourbier commented on pymialsrtk/interfaces/preprocess.py:
```diff
@@ -155,14 +174,29 @@ class MialsrtkCorrectSliceIntensityInputSpec(BaseInterfaceInputSpec):
     in_file = File(desc='Input image', mandatory=True)
     in_mask = File(desc='Input mask', mandatory=False)
-    out_postfix = traits.Str("", usedefault=True)
+    out_postfix = traits.Str("", desc='Suffixe to be added to in_file', usedefault=True) 
```
Not sure is there is not a white space inserted here

sebastientourbier reviewed (commented):
@hamzake On the right way. We still need to define more appropriately the references. 

@hamzake @pdedumast Here is some precisions regarding how to include references in the docstring. For each docstring where there is a citation, the citation will appear as [1]_. If there are multiple they will appear as [1]_, [2]_, ... [N]_ in the order they appear. And then the citations will be define in the References section. Here is an example with the histogram normalization module:
```
Runs the MIAL SRTK histogram normalizaton module proposed by Nyúl et al. [1]_.

References
------------
.. [1] Nyúl et al.; Medical Imaging, IEEE Transactions, 2000. `(link to paper)<https://ieeexplore.ieee.org/document/836373>`_

Example
----------
>>> from pymialsrtk.interfaces.preprocess import MialsrtkHistogramNormalization
>>> histNorm = MialsrtkHistogramNormalization()
>>> histNorm.inputs.bids_dir = '/my_directory'
>>> histNorm.inputs.input_images = ['image1.nii.gz','image2.nii.gz']
>>> histNorm.inputs.input_masks = ['mask1.nii.gz','mask2.nii.gz']
>>> histNorm.inputs.out_postfix = '_histnorm'
>>> histNorm.inputs.stacksOrder = 
>>> histNorm.run()  # doctest: +SKIP
```
Note that the numbering of the references is relative to each docstring and the references should be described in each docstring. 
Note also the `# doctest: +SKIP` as we do not want this to be executed as a test

Here is a good docstring template from dipy that I followed:

https://github.com/dipy/dipy/blob/8883aa761f62a5bbd800c893ff40ffc4b653ea38/dipy/reconst/dsi.py#L19-L105

sebastientourbier commented:
Here is a good docstring template for functions:
https://github.com/dipy/dipy/blob/8883aa761f62a5bbd800c893ff40ffc4b653ea38/dipy/reconst/dsi.py#L201-L230

sebastientourbier reviewed (commented):
References look good

hamzake commented:
> Here is a good docstring template for functions:
> https://github.com/dipy/dipy/blob/8883aa761f62a5bbd800c893ff40ffc4b653ea38/dipy/reconst/dsi.py#L201-L230

Thanks, I'll use it to document the brain extraction part

sebastientourbier commented:
@pdedumast @hamzake Seems it is going very well ! Just to let you know I incorporated the generation of the doc from the docstring. I am still working on some issues in the `documentation/conf.py` with packages not listed in `documentation/requirements.txt` and so not installed in readthedocs when it builds the doc. But these should not conflict with your changes. But to make sure you are always up-to-date with this branch, I advice you to always do `git pull` before you commit and push.

sebastientourbier commented:
Existing docstrings are now correctly rendered.  
Page for the preprocess module: https://mialsrtk.readthedocs.io/en/dev-pgd-hk/api_preprocess.html

sebastientourbier commented:
@pdedumast @hamzake This starts to really look great! Thanks a lot for all this work already!

By looking at the rendered doc, I guess the parameter you are describing are more the attributes of the InputSpec classes. In fact, I would suggest to also add a docstring for the Input/OutputSpec classes where we move the definition of these parameters. It would be great to make a link to their interface as the following:
```
See Also
----------

pymialsrtk.interfaces.preprocess.srtkHistogramNormalization
```

Regarding the pipeline, it seems we should review the way to write the example. I guess we should just follow the way it is done for the interface. Also, it will be helpful to have also a description of the class attributes.
Also be careful with code style, codacy is not really happy at the moment ;)

hamzake commented:
Hey Seb,

Something like this ?

class BtkNLMDenoisingInputSpec(BaseInterfaceInputSpec):
    bids_dir = Directory(desc='BIDS root directory', mandatory=True, exists=True)
    in_file = File(desc='Input image filename', mandatory=True)
    in_mask = File(desc='Input mask filename', mandatory=False)
    out_postfix = traits.Str("_nlm", desc='Suffix to be added to input image filename to construst denoised output filename', usedefault=True)
    weight = traits.Float(0.1, desc='NLM smoothing parameter (0.1 by default)', usedefault=True)
    
    """Parameters
    ----------
    bids_dir <string>
        BIDS root directory (required)

    in_file <string>
        Input image file (required)

    in_mask <string>
        Mask of the input image

    out_postfix <string>
        suffix added to input image filename to construct output filename (default is '_nlm')

    weight <float>
        smoothing parameter (high beta produces smoother result, default is 0.1)
    class BtkNLMDenoisingOutputSpec(TraitedSpec):
    out_file = File(desc='Output denoised image’)
    """
    See also
    --------------
    pymialsrtk.interfaces.preprocess.BtkNLMDenoising    

class BtkNLMDenoising(BaseInterface):
    """
    Runs the non-local mean denoising module: implementation by Rousseau et al. [1]_ of the method proposed by Coupé et al. [2]_.

    References
    -----------
    .. [1] Rousseau et al.; Computer Methods and Programs in Biomedicine, 2013. `(link to paper) <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3508300>`_
    .. [2] Coupé et al.; IEEE Transactions on Medical Imaging, 2008. `(link to paper) <https://doi.org/10.1109/tmi.2007.906087>`_

    Example
    ---------
    >>> from pymialsrtk.interfaces.preprocess import BtkNLMDenoising
    >>> nlmDenoise = BtkNLMDenoising()
    >>> nlmDenoise.inputs.bids_dir = '/my_directory'
    >>> nlmDenoise.inputs.in_file = 'my_image.nii.gz'
    >>> nlmDenoise.inputs.in_mask = 'my_mask.nii.gz'
    >>> nlmDenoise.inputs.weight = 0.2
    >>> nlmDenoise.run() # doctest: +SKIP
    """

> Le 30 oct. 2020 à 16:06, Sébastien Tourbier <notifications@github.com> a écrit :
> 
> 
> @pdedumast <https://github.com/pdedumast> @hamzake <https://github.com/hamzake> This starts to really look great! Thanks a lot for all this work already!
> 
> By looking at the rendered doc, I guess the parameter you are describing are more the attributes of the InputSpec classes. In fact, I would suggest to also add a docstring for the Input/OutputSpec classes where we move the definition of these parameters. It would be great to make a link to their interface as the following:
> 
> See Also
> ----------
> 
> pymialsrtk.interfaces.preprocess.srtkHistogramNormalization
> Regarding the pipeline, it seems we should review the way to write the example. I guess we should just follow the way it is done for the interface. Also, it will be helpful to have also a description of the class attributes.
> Also be careful with code style, codacy is not really happy at the moment ;)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/pull/2#issuecomment-719607522>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGTMTDQXFKH6CTPKE7S7BYLSNLI6LANCNFSM4NQRXVAQ>.
>

sebastientourbier commented:
I would say something like:

```python
class BtkNLMDenoisingInputSpec(BaseInterfaceInputSpec):
    """Class used to represent inputs of the BtkNLMDenoising interface.

    Attributes
    ----------
    bids_dir <string>
        BIDS root directory (required)

    in_file <string>
        Input image file (required)

    in_mask <string>
        Mask of the input image

    out_postfix <string>
        suffix added to input image filename to construct output filename (default is '_nlm')

    weight <float>
        smoothing parameter (high beta produces smoother result, default is 0.1)

    See Also
    ----------
    pymialsrtk.interfaces.preprocess.BtkNLMDenoising
    """

    bids_dir = Directory(desc='BIDS root directory', mandatory=True, exists=True)
    in_file = File(desc='Input image filename', mandatory=True)
    in_mask = File(desc='Input mask filename', mandatory=False)
    out_postfix = traits.Str("_nlm", desc='Suffix to be added to input image filename to construst denoised output filename', usedefault=True)
    weight = traits.Float(0.1, desc='NLM smoothing parameter (0.1 by default)', usedefault=True)


class BtkNLMDenoisingOutputSpec(TraitedSpec):
    """Class used to represent outputs of the BtkNLMDenoising interface.
    
    Attributes
    -----------   
    out_file <string>
        Output denoised image file 

    See also
    --------------
    pymialsrtk.interfaces.preprocess.BtkNLMDenoising
    """

    out_file = File(desc='Output denoised image’)
     
 
class BtkNLMDenoising(BaseInterface):
    """
    Runs the non-local mean denoising module: implementation by Rousseau et al. [1]_ of the method proposed by Coupé et al. [2]_.

    References
    -----------
    .. [1] Rousseau et al.; Computer Methods and Programs in Biomedicine, 2013. `(link to paper) <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3508300>`_
    .. [2] Coupé et al.; IEEE Transactions on Medical Imaging, 2008. `(link to paper) <https://doi.org/10.1109/tmi.2007.906087>`_

    Example
    ---------
    >>> from pymialsrtk.interfaces.preprocess import BtkNLMDenoising
    >>> nlmDenoise = BtkNLMDenoising()
    >>> nlmDenoise.inputs.bids_dir = '/my_directory'
    >>> nlmDenoise.inputs.in_file = 'my_image.nii.gz'
    >>> nlmDenoise.inputs.in_mask = 'my_mask.nii.gz'
    >>> nlmDenoise.inputs.weight = 0.2
    >>> nlmDenoise.run() # doctest: +SKIP
    """

```

sebastientourbier commented:
@hamzake Seems since commit e1456c1d8608d5a830256a03458fa2a9ddd8f4d7, the test run is failing on circleci. By looking in more details at the outputs (here: https://app.circleci.com/pipelines/github/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/149/workflows/2ba720ea-7450-4fe8-92c9-0331d0cb1a68/jobs/224/parallel-runs/0/steps/0-106) it seems it is related to the change in the naming of the variable stackOrder in the code for the interface but there is still a need to change it in the srr.py where it is called.

sebastientourbier reviewed (commented):
Great! Let see if circleci is now happy 🤞

sebastientourbier commented:
@hamzake  By looking more carefully at the test that ran on circleci, we still have some issues with the following error:
```bash
201102-10:38:39,174 nipype.interface INFO:
	 **** Processing ****
Traceback (most recent call last):
  File "/opt/mialsuperresolutiontoolkit/docker/bidsapp/run.py", line 82, in <module>
    use_manual_masks=args.manual)
  File "/opt/mialsuperresolutiontoolkit/docker/bidsapp/run.py", line 35, in main
    pipeline.create_workflow()
  File "/opt/conda/envs/pymialsrtk-env/lib/python3.6/site-packages/pymialsrtk/pipelines/anatomical/srr.py", line 245, in create_workflow
    srtkImageReconstruction.inputs.stacks_order = self.p_stacks_order
traits.trait_errors.TraitError: Cannot set the undefined 'stacks_order' attribute of a 'MialsrtkImageReconstructionInputSpec' object.
```

sebastientourbier commented:
Seems stacksOrder still appearing in postprocess:
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/2a0a31354ecf02df1034b846391667d260dceaaf/pymialsrtk/interfaces/postprocess.py#L26-L169

sebastientourbier commented:
@pdedumast @hamzake Just to let you know I corrected the output structure, updated the list of outputs to make circleci happy and improved the documentation (in particular, I rewrote the about section in the ``index.rst``, made big modifications in ``outputs.rst`` and updated the README. Now all tests pass on GitHub.

Please could you have a review of the doc a last time? 
Then if you do not have any major concern, I would say we are ready to merge 😎 
There are still a few little things that we will have to do before the release for which I created individual issue (#22 #23 #24) as this PR has become giant. If you see others that I may have forgotten please feel free to create them.

Thanks for all this huge work Priscille and Hamza!!!🎉

sebastientourbier commented:
@all-contributors please add @hamzake for doc

sebastientourbier commented:
@all-contributors please add @pdedumast for doc
